lib/pengine/unpack.c

```diff
@@ -1486,7 +1491,15 @@ determine_online_status_fencing(pe_working_set_t *data_set,
 
     } else if (do_terminate == FALSE && pcmk__str_eq(exp_state, CRMD_JOINSTATE_DOWN, pcmk__str_casei)) {
 
-        if (member || crmd_online) {
+        if (when_member > 0
+            && when_online == 0
+            && (get_effective_time(data_set) - when_member
+                >= data_set->node_pending_timeout)) {
+            pe_fence_node(data_set, this_node,
+                          "peer pending timed out on joining the process group",
+                          FALSE);
+
+        } else if (member || crmd_online) {
```

Comment:
We have two problematic scenarios with this:

* If the node is leaving instead of joining, and the scheduler runs after it has left the process group but before it has left corosync, it will likely be fenced.
* If the user intentionally starts only corosync on the node (for quorum purposes, troubleshooting, etc.), the node will be fenced

I'm thinking of relying entirely on the controller instead of the scheduler changes here. The controller would start the timer when the node begins the join process rather than when it joins the cluster (which eliminates both the scenarios above), and if the timer pops, it updates the node_state in a way that the scheduler sees as unclean.

Thoughts?

Reply:
@gao-yan My memory is hazy -- why were join-integration-timeout and join-finalization-timeout not enough?

Reply:
> @gao-yan My memory is hazy -- why were join-integration-timeout and join-finalization-timeout not enough?

I think they are for the situations where a node has already joined CPG and is doing actual "integration"?

Reply:
> Maybe whenever we update crmd to offline, we also update in_ccm to true instead of a timestamp? That would skip the pending timeout code. We wouldn't want to overwrite in_ccm if it was 0/false, but hopefully there's a way of knowing that at that point.

Unfortunately `peer_update_callback()` can call `create_node_state_update()` with `node_update_peer` and without `node_update_cluster`. (And it's important that it doesn't update `node_update_cluster`.) All other callers set `node_update_cluster` anytime they set `node_update_peer`.

Maybe `peer_update_callback()` could update in_ccm to true when called for losing the CPG flag and the peer's cluster flag is still set.

Reply:
Sounds like a good idea.